src/python/pants/backend/python/lint/yapf/subsystem.py

```diff
@@ -0,0 +1,105 @@
+from pants.option.custom_types import file_option, shell_str
+
+
+class Yapf(PythonToolBase):
+    options_scope = "yapf"
+    help = "A formatter for Python files (https://github.com/google/yapf)."
+
+    default_version = "yapf==0.31.0"
+    default_extra_requirements = ["setuptools", "toml"]
+    register_interpreter_constraints = True
+    default_interpreter_constraints = ["CPython>=3.6"]
+    default_main = ConsoleScript("yapf")
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--skip",
+            type=bool,
+            default=False,
+            help=(
+                f"Don't use yapf when running `{register.bootstrap.pants_bin_name} fmt` and "
+                f"`{register.bootstrap.pants_bin_name} lint`."
+            ),
+        )
+        register(
+            "--args",
+            type=list,
+            member_type=shell_str,
+            help=(
+                "Arguments to pass directly to yapf, e.g. "
+                f'`--{cls.options_scope}-args="--no-local-style"`. '
+                "Certain arguments, specifically `--recursive`, `--in-place`, and "
+                "`--parallel`, will be ignored because Pants takes care of finding "
+                "all the relevant files and running the formatting in parallel."
+            ),
+        )
+        register(
+            "--config",
+            type=file_option,
+            default=None,
+            advanced=True,
+            help=(
+                "Path to style file understood by yapf "
+                "(https://github.com/google/yapf#formatting-style/).\n\n"
+                f"Setting this option will disable `[{cls.options_scope}].config_discovery`. Use "
+                f"this option if the config is located in a non-standard location."
+            ),
+        )
+        register(
+            "--config-discovery",
+            type=bool,
+            default=True,
+            advanced=True,
+            help=(
+                "If true, Pants will include any relevant config files during "
+                "runs (`.style.yapf`, `pyproject.toml`, `setup.cfg`, and `~/.config/yapf/style`)."
+                f"\n\nUse `[{cls.options_scope}].config` instead if your config is in a "
+                f"non-standard location."
+            ),
+        )
+
+    @property
+    def skip(self) -> bool:
+        return cast(bool, self.options.skip)
+
+    @property
+    def args(self) -> tuple[str, ...]:
+        return tuple(self.options.args)
+
+    @property
+    def config(self) -> str | None:
+        return cast("str | None", self.options.config)
+
+    def config_request(self, dirs: Iterable[str]) -> ConfigFilesRequest:
+        # Refer to https://github.com/google/yapf#formatting-style.
+        check_content = {}
+        for d in ("", *dirs):
+            check_content.update(
+                {
+                    os.path.join(d, "pyproject.toml"): b"[tool.yapf]",
```

Comment:
This is more robust to if you have an entry like `[tool.yapf.reports]`.

```suggestion
                    os.path.join(d, "pyproject.toml"): b"[tool.yapf",
```